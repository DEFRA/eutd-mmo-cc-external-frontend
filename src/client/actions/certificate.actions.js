import { isError, showFullPageError } from './index';

export const GET_SUMMARY_DOCUMENT = 'GET_SUMMARY_DOCUMENT';
export const SHOW_SUMMARY_DOCUMENT_ERRORS = 'SHOW_SUMMARY_DOCUMENT_ERRORS';
export const GET_SUMMARY_UNAUTHORISED = 'GET_SUMMARY_DOCUMENT_UNAUTHORISED';

export function showSummaryCertificateErrors(data) {
  return { type: SHOW_SUMMARY_DOCUMENT_ERRORS, payload: { validationErrors: data } };
}

export const getSummaryCertificate = (journey, documentNumber = undefined) => async (dispatch, getState, { orchestrationApi }) => {
  const config = (documentNumber)
    ? { headers: { documentnumber: documentNumber } }
    : {};

  await orchestrationApi.get(`/certificate/${journey}`, config)
    .then(
      res => {
        dispatch({ type: GET_SUMMARY_DOCUMENT, payload: res.data });
      },
      error => {
        if (isError(error)) {
          dispatch(showFullPageError());
        } else {
          dispatch(unauthorised());
        }
      });

    function unauthorised() { return { type: GET_SUMMARY_UNAUTHORISED }; }
};