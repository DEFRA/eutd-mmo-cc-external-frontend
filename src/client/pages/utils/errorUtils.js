import i18n from '../../../i18n';
import enTranslations from '../../../locales/uk/en.json';

export function toGovukErrors(obj) {
  return Object.keys( obj).map( o => {
    return {targetName: o, text: obj[o], key: o, message: (obj[o] in enTranslations) ? i18n.t(obj[o]) : obj[o]};
  });
}

export function onHandleErrorClick(targetName, speciesRow, vesselIndex) {
  let elem;

  if (speciesRow !== undefined && vesselIndex !== undefined) {
    elem = document.getElementById(targetName + '-' + speciesRow + '-' + vesselIndex);
  } else {
    elem = document.getElementsByName(targetName)[0] || document.getElementById(targetName);
  }

  if (elem) {
    elem.scrollIntoView({ behavior: 'smooth' });
    elem.focus({ preventScroll: true});
  }
}

export function onHandleErrorClickWithoutFocus(targetName) {
  const elem = document.getElementsByName(targetName)[0] || document.getElementById(targetName);
  elem.scrollIntoView({ behavior: 'smooth' });
}

export function scrollToErrorIsland() {
  setTimeout( () => {
    document.getElementById('errorIsland').scrollIntoView({behavior: 'smooth'});
    document.getElementById('errorIsland').focus({ preventScroll: true});
  },100);
}

export function scrollToField(id) {
  setTimeout( () => {
    document.getElementById(id).scrollIntoView({behavior: 'smooth'});
    document.getElementById(id).focus({ preventScroll: true});
  },100);
}

export function scrollToFieldName(name) {
  setTimeout( () => {
    document.getElementsByName(name)[0].scrollIntoView({behavior: 'smooth'});
    document.getElementsByName(name)[0].focus({ preventScroll: true});
  },100);
}

