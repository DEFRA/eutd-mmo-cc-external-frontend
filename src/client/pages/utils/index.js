import React, {Fragment} from 'react';

export * from './errorUtils';


export function formatAddress( ...adresssLines) {
  return adresssLines.filter( a => a ).map( (a, index) => <Fragment key={index}>{a}<br/></Fragment> );
}

export function constructAddress(addressArray) {
  let address = '';
  if (addressArray.length === 4) {
    if (addressArray[0]) {
      address += addressArray[0];
    }
    if (addressArray[0] && addressArray[1]) {
      address += ' ';
    }
    if (addressArray[1]) {
      address += addressArray[1];
    }
    if (addressArray[2]) {
      address += ', ' + addressArray[2];
    }
    if (addressArray[3]) {
      address += '. ' + addressArray[3];
    }
  }
  return address;
}

export const hasAdminOverride = (exportPayload) => {
  const isLandingOverridden = (landing) =>
    landing &&
    landing.model &&
    landing.model.vessel &&
    landing.model.vessel.vesselOverriddenByAdmin;

  return exportPayload &&
    exportPayload.items &&
    exportPayload.items.some(
      (item) =>
        item.landings &&
        item.landings.some((landing) => isLandingOverridden(landing))
    )
    ? true
    : false;
};