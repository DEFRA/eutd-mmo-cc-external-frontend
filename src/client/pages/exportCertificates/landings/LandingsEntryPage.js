import React, { useEffect, useState } from 'react';
import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import { isEmpty } from 'lodash';
import { BackLink, GridCol, GridRow, H1, Main, MultiChoice } from 'govuk-react';
import { onHandleErrorClick, scrollToErrorIsland } from '../../utils';
import { getLandingType, validateLandingType, clearLandingsType, dispatchSaveChangedLandingsType } from '../../../actions/landingsType.actions';
import ContinueButton from '../../../components/elements/ContinueButton';
import Details from '../../../components/elements/Details';
import ErrorIsland from '../../../components/elements/ErrorIsland';
import Form from '../../../components/elements/Form';
import HelpLink from '../../../components/HelpLink';
import PageTemplateWrapper from '../../../components/PageTemplateWrapper';
import PageTitle from '../../../components/PageTitle';
import SelectRadio from '../../../components/elements/SelectRadio';
import NotificationBanner from '../../../components/NotificationBanner';
import { withTranslation } from 'react-i18next';
import { t } from 'i18next';

const LandingsEntryPage = (props) => {
  const pageTitle = t('ccLandingsEntryPageTitle');
  const {
    previousUri,
    path,
    nextUri,
    landingsEntryOptions,
    journey,
    landingsTypeConfirmationUri,
    title
  } = props.route;
  const { params } = props.match;
  const confirmCopyDocument = props.confirmCopyDocument;
  const { documentNumber } = params;

  const [landingsEntryValue, setLandingsEntryValue] = useState(props.landingsEntryOption);

  useEffect(() => {
    window.scrollTo(0, 0);
    props.getLandingType(documentNumber);

    return () => {
      props.clearLandingsType();
    };
  }, []);

  useEffect(() => {
    if (props.unauthorised)
      props.history.push('/forbidden');

    setLandingsEntryValue(props.landingsEntryOption);
  }, [props.landingsEntryOption]);

  const handleBackLinkClick = (event) => {
    event.preventDefault();
    props.history.push(previousUri.replace(':documentNumber', documentNumber));
  };

  const handleLandingsOptionChange = (event) => {
    setLandingsEntryValue(event.currentTarget.value);
  };

  const getNotificationMsg = (generatedByContent) => {
    let notificationMsgs = [];
     if(confirmCopyDocument && confirmCopyDocument.copyDocumentAcknowledged)
       notificationMsgs.push(
         confirmCopyDocument.voidDocumentConfirm ?
         t('ccProgressNotificationMsgIsVoided') :
         t('ccProgressNotificationMsgIsNotVoided', { documentNumber : confirmCopyDocument.documentNumber })
       );
    if(generatedByContent)
      notificationMsgs.push(t('ccLandingsEntryPageNewPageMessage'));

    return notificationMsgs.length ? notificationMsgs : null;
  };

  const handleSaveAndContinue = async (event) => {
    event.preventDefault();
    try {
      const hasChangedLandingType = () =>
        props.landingsEntryOption !== null &&
        (landingsEntryValue === 'directLanding' || props.landingsEntryOption === 'directLanding')
        && props.landingsEntryOption !== landingsEntryValue;

      if (hasChangedLandingType()) {
        await props.dispatchSaveChangedLandingsType(landingsEntryValue);
        props.history.push(landingsTypeConfirmationUri.replace(':documentNumber', documentNumber));
      } else {
        await props.validateLandingType(landingsEntryValue, documentNumber);
        props.history.push(nextUri.replace(':documentNumber', documentNumber));
      }
    } catch {
      scrollToErrorIsland();
    }
  };

  const renderLandingsEntryOptions = () => {
    return (
      Array.isArray(landingsEntryOptions) &&
      landingsEntryOptions.map((option) => (
        <SelectRadio
          key={option.id}
          id={option.id}
          name={option.name}
          value={option.value}
          hint={t(option.hint)}
          onChange={handleLandingsOptionChange}
          checked={landingsEntryValue === option.value}
        >
          {t(option.label)}
        </SelectRadio>
      ))
    );
  };

  return (
    <Main>
      {!isEmpty(props.errors) && (
        <ErrorIsland
          errors={props.errors.errors.map((err) => ({
            message: t(err.text),
            key: err.targetName,
          }))}
          onHandleErrorClick={onHandleErrorClick}
        />
      )}
      <PageTitle
        title={`${
          !isEmpty(props.errors) ? 'Error: ' : ''
        }${pageTitle} - ${t(title)}`}
      />
      <Form
        action='/v1/export-certificates/landings-type'
        currentUrl={path.replace(':documentNumber', documentNumber)}
        nextUrl={nextUri.replace(':documentNumber', documentNumber)}
        documentNumber={documentNumber}
      >
        <GridRow>
          <GridCol>
            <BackLink onClick={handleBackLinkClick} href={previousUri.replace(':documentNumber', documentNumber)} >{t('commonBackLinkBackButtonLabel')}</BackLink>
            {(props.generatedByContent || ( confirmCopyDocument && confirmCopyDocument.copyDocumentAcknowledged)) && (
              <NotificationBanner
                header={t('commonImportant')}
                messages={getNotificationMsg(props.generatedByContent)} />
            )}
            <H1>{pageTitle}</H1>
            <MultiChoice
              meta={{
                touched: true,
                error:
                  !isEmpty(props.errors) &&
                  t(props.errors.landingsEntryOptionError),
              }}
              id="landingsEntryOption"
            >
              <fieldset style={{ marginBottom: '30px'}}>{renderLandingsEntryOptions()}</fieldset>
            </MultiChoice>
            <Details
              summary={t('ccLandingsEntryPageWhatIsACSVFile')}
              details={
                <>
                  <span>{t('ccLandingsEntryPageWhatIsACSVFileFormatDetails')}</span>
                  <br />
                  <br />
                  <span>{t('ccLandingsEntryPageWhatIsACSVFileExportingDetails')}</span>
                </>
              }
            />
          </GridCol>
        </GridRow>
        <GridRow>
            <ContinueButton id="continue" onClick={handleSaveAndContinue}>{t('commonContinueButtonSaveAndContinueButton')}</ContinueButton>
          </GridRow>
      </Form>
      <HelpLink journey={journey} parent="functionalComp"/>
    </Main>
  );
};

function mapStateToProps(state) {
  return {
    errors: state.errors,
    landingsEntryOption: state.landingsType.landingsEntryOption,
    generatedByContent: state.landingsType.generatedByContent,
    unauthorised: state.landingsType.unauthorised,
    confirmCopyDocument: state.confirmCopyDocument
  };
}

function loadData(store) {
  store.dispatch(getLandingType(this.documentNumber));
}

export const component = withRouter(
  connect(mapStateToProps, {
    getLandingType,
    validateLandingType,
    clearLandingsType,
    dispatchSaveChangedLandingsType,
  })(withTranslation()(LandingsEntryPage))
);

export default {
  loadData,
  component: PageTemplateWrapper(component),
};
