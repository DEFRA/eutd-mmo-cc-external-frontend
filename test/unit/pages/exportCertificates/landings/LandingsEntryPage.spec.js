import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter, Router } from 'react-router';
import { Route, Switch } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import MutationObserver from 'mutation-observer';
import { component as LandingsEntryPage } from '../../../../../src/client/pages/exportCertificates/landings/LandingsEntryPage';
import LandingsEntryPageExport from '../../../../../src/client/pages/exportCertificates/landings/LandingsEntryPage';
import { getLandingType, validateLandingType, dispatchSaveChangedLandingsType, clearLandingsType} from '../../../../../src/client/actions/landingsType.actions';

import * as Utils from '../../../../../src/client/pages/utils/errorUtils';


jest.mock('../../../../../src/client/actions/landingsType.actions');

const renderComponent = (state) => {

  const store = configureStore()(state);

  const props = {
    route: {
      title: 'Create a UK catch certificate - GOV.UK',
      previousUri: '/create-catch-certificate/:documentNumber/add-exporter-details',
      nextUri: '/create-catch-certificate/:documentNumber/progress',
      path: '/create-catch-certificate/:documentNumber/landings-entry',
      uploadFileUri: '/create-catch-certificate/:documentNumber/upload-file',
      landingsTypeConfirmationUri: '/create-catch-certificate/:documentNumber/landings-type-confirmation',
      journey: 'catchCertificate',
      landingsEntryOptions: [
        {
          label: 'Direct Landing',
          value: 'directLanding',
          name: 'landingsEntry',
          id: 'directLandingOptionEntry',
          hint: 'Recommended for UK registered fishing vessels landing and exporting their catch simultaneously in the EU (or a GB registered fishing vessel direct landing in Northern Ireland).',
        },
        {
          label: 'Manual entry',
          value: 'manualEntry',
          name: 'landingsEntry',
          id: 'manualOptionEntry',
          hint: 'Recommended for small to medium sized exports.',
        },
        {
          label: 'Upload from a CSV file',
          value: 'uploadEntry',
          name: 'landingsEntry',
          id: 'uploadOptionEntry',
          hint: 'Recommended for large exports. (Requires the set up of product favourites).',
        }
      ]
    }
  };

  const history = createMemoryHistory({ initialEntries: ['document123/landings-entry'] });

  const { container } = render(
    <Router history={history}>
      <Switch>
        <Route exact path="/create-catch-certificate/:documentNumber/add-exporter-details">
           <div>Exporter Detaiils Page</div>
        </Route>
        <Route exact path="/create-catch-certificate/catch-certificates">
           <div>Exporter Dashboard Page</div>
        </Route>
        <Route exact path="/create-catch-certificate/:documentNumber/landings-type-confirmation">
           <div>Confirm Landings Type Page</div>
        </Route>
        <Route exact path="/create-catch-certificate/:documentNumber/progress">
           <div>Catch Certificate application</div>
        </Route>
        <Route exact path="/forbidden">
           <div>Forbidden</div>
        </Route>
        <Route exact path=":documentNumber/landings-entry">
          <Provider store={store}>
              <LandingsEntryPage {...props} />
          </Provider>
        </Route>
      </Switch>
    </Router>
  );

  return container;
};

beforeEach(() => {
  getLandingType.mockReturnValue({ type: 'GET_LANDINGS_TYPE' });
});

afterEach(() => {
  getLandingType.mockReset();
});

describe('LandingsEntryPage', () => {
  const defaultStore = {
    config: {
      catchCertHelpUrl: 'https://www.gov.uk/guidance/exporting-and-importing-fish-if-theres-no-brexit-deal'
    },
    errors: {},
    landingsType: {
      landingsEntryOption: 'directLanding',
      generatedByContent: false
    }
  };

  let mockScrollToErrorIsland;
  let mockWindowScrollTo;
  let originalMutationObserver;

  beforeEach(() => {
    validateLandingType.mockReturnValue({ type: 'LANDINGS_TYPE' });
    dispatchSaveChangedLandingsType.mockReturnValue({ type: 'save_changed_landings_type' });
    clearLandingsType.mockReturnValue({tyep:'landingsType/landings_type/clear'});
    mockScrollToErrorIsland = jest.spyOn(Utils, 'scrollToErrorIsland');
    mockScrollToErrorIsland.mockReturnValue(undefined);
    mockWindowScrollTo = jest.fn()
      .mockReturnValue(undefined);

    window.scrollTo = mockWindowScrollTo;

    originalMutationObserver = global.MutationObserver;
    global.MutationObserver = MutationObserver;
  });

  afterEach(() => {
    mockScrollToErrorIsland.mockRestore();
    mockWindowScrollTo.mockRestore();
    validateLandingType.mockRestore();
    dispatchSaveChangedLandingsType.mockRestore();

    global.MutationObserver = originalMutationObserver;
  });

  it('should render an LandingsEntry component', () => {
    const wrapper = renderComponent(defaultStore);
    expect(wrapper).toBeDefined();
  });

  it('should scroll to top', () => {
    renderComponent(defaultStore);
    expect(mockWindowScrollTo).toHaveBeenCalled();
  });

  it('should get saved landings entry option when the page loads', () => {
    renderComponent(defaultStore);
    expect(getLandingType).toHaveBeenCalled();
    expect(getLandingType).toHaveBeenCalledWith('document123');
    expect(getLandingType).toHaveBeenCalledTimes(1);
  });

  it('should clear the landings entry type', () => {
    expect(clearLandingsType).toHaveBeenCalledWith();
  });

  it('should select previous selection on landing entry page', () => {
    const wrapper = renderComponent(defaultStore);
    const directLanding = wrapper.querySelector('#directLandingOptionEntry');
    const manualEntry = wrapper.querySelector('#manualOptionEntry');
    const uploadEntry = wrapper.querySelector('#uploadOptionEntry');

    expect(directLanding.checked).toBeTruthy();
    expect(manualEntry.checked).toBeFalsy();
    expect(uploadEntry.checked).toBeFalsy();
  });

  it('should render a form element', () => {
    const wrapper = renderComponent(defaultStore);
    const form = wrapper.querySelector('form');
    expect(form).toBeDefined();
  });

  it('should render a back link component', () => {
    renderComponent(defaultStore);
    const backLink = screen.getByText('Back');
    expect(backLink).toHaveAttribute('href', '/create-catch-certificate/document123/add-exporter-details');
  });

  it('should render a page title', () => {
    renderComponent(defaultStore);
    expect(screen.getByText('How do you want to enter your products and landings?')).toBeDefined();
  });

  it('should render all landings entry options', () => {
    renderComponent(defaultStore);
    expect(screen.getByText('Direct Landing')).toBeDefined();
    expect(screen.getByText('Manual entry')).toBeDefined();
    expect(screen.getByText('Upload from a CSV file')).toBeDefined();
  });

  it('should render all landings entry options with hints', () => {
    renderComponent(defaultStore);
    expect(screen.getByText('Recommended for UK registered fishing vessels landing and exporting their catch simultaneously in the EU (or a GB registered fishing vessel direct landing in Northern Ireland).')).toBeDefined();
    expect(screen.getByText('Recommended for small to medium sized exports.')).toBeDefined();
    expect(screen.getByText('Recommended for large exports. (Requires the set up of product favourites).')).toBeDefined();
  });

  it('should select the direct landing radio option', () => {
    const wrapper = renderComponent(defaultStore);
    const directLanding = wrapper.querySelector('#directLandingOptionEntry');
    const manualEntry = wrapper.querySelector('#manualOptionEntry');
    const uploadEntry = wrapper.querySelector('#uploadOptionEntry');

    fireEvent(directLanding, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(directLanding.checked).toBeTruthy();
    expect(manualEntry.checked).toBeFalsy();
    expect(uploadEntry.checked).toBeFalsy();
  });

  it('should select the manual entry radio option', () => {
    const wrapper = renderComponent(defaultStore);
    const directLanding = wrapper.querySelector('#directLandingOptionEntry');
    const manualEntry = wrapper.querySelector('#manualOptionEntry');
    const uploadEntry = wrapper.querySelector('#uploadOptionEntry');

    fireEvent(manualEntry, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(manualEntry.checked).toBeTruthy();
    expect(directLanding.checked).toBeFalsy();
    expect(uploadEntry.checked).toBeFalsy();
  });

  it('should select the upload entry radio option', () => {
    const wrapper = renderComponent(defaultStore);
    const directLanding = wrapper.querySelector('#directLandingOptionEntry');
    const manualEntry = wrapper.querySelector('#manualOptionEntry');
    const uploadEntry = wrapper.querySelector('#uploadOptionEntry');

    fireEvent(uploadEntry, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(uploadEntry.checked).toBeTruthy();
    expect(manualEntry.checked).toBeFalsy();
    expect(directLanding.checked).toBeFalsy();
  });

  it('should have a details component', () => {
    renderComponent(defaultStore);
    expect(screen.getByText('What is a CSV file?')).toBeDefined();
    expect(screen.getByText('A CSV file is a text file that uses commas to separate its values. Each value in the file is a data field and each line is a data record.')).toBeDefined();
    expect(screen.getByText('Spreadsheets and some software can usually export their data to a CSV file.')).toBeDefined();
  });

  it('should have a Save and continue button', () => {
    renderComponent(defaultStore);
    expect(screen.getByText('Save and continue')).toBeDefined();
  });

  it('should call the Save and continue button handler and navigate to the progress page', async () => {
    const wrapper = renderComponent(defaultStore);
    const saveAndContinueBtn = screen.getByText('Save and continue');
    const directLanding = wrapper.querySelector('#directLandingOptionEntry');

    fireEvent(directLanding, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    fireEvent(saveAndContinueBtn, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(validateLandingType).toHaveBeenCalledWith('directLanding', 'document123');
    expect(await screen.findByText('Catch Certificate application')).toBeDefined();
  });

  it('should call the Save and continue button handler and navigate to the landings confirmation uri', async () => {
    const wrapper = renderComponent(defaultStore);
    const saveAndContinueBtn = screen.getByText('Save and continue');
    const uploadFile = wrapper.querySelector('#uploadOptionEntry');

    fireEvent(uploadFile, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    fireEvent(saveAndContinueBtn, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(dispatchSaveChangedLandingsType).toHaveBeenCalledWith('uploadEntry');
    expect(await screen.findByText('Confirm Landings Type Page')).toBeDefined();
  });

  it('should call the Save and continue button handler and navigate to the progress page when landing type is initally undefined', async () => {
    const wrapper = renderComponent({
      config: {
        catchCertHelpUrl: 'https://www.gov.uk/guidance/exporting-and-importing-fish-if-theres-no-brexit-deal'
      },
      errors: {},
      landingsType: {
        landingsEntryOption: null,
        generatedByContent: false
      }
    });
    const saveAndContinueBtn = screen.getByText('Save and continue');
    const directLanding = wrapper.querySelector('#directLandingOptionEntry');

    fireEvent(directLanding, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    fireEvent(saveAndContinueBtn, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(validateLandingType).toHaveBeenCalledWith('directLanding', 'document123');
    expect(await screen.findByText('Catch Certificate application')).toBeDefined();
  });

  it('should call the Save and continue button handler and navigate to the progress page when landings entry is upload entry', async () => {
    renderComponent({
      config: {
        catchCertHelpUrl: 'https://www.gov.uk/guidance/exporting-and-importing-fish-if-theres-no-brexit-deal'
      },
      errors: {},
      landingsType: {
        landingsEntryOption: 'uploadEntry',
        generatedByContent: false
      }
    });

    const saveAndContinueBtn = screen.getByText('Save and continue');

    fireEvent(saveAndContinueBtn, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(validateLandingType).toHaveBeenCalledWith('uploadEntry', 'document123');
    expect(await screen.findByText('Catch Certificate application')).toBeDefined();
  });

  it('should call the Save and continue button handler and catch any thrown errors', () => {
    renderComponent(defaultStore);
    validateLandingType.mockImplementation(() => {
      throw new Error('Bad Request');
    });

    const saveAndContinueBtn = screen.getByText('Save and continue');

    fireEvent(saveAndContinueBtn, new MouseEvent('click', {
      bubbles: true,
      cancelable: true
    }));

    expect(mockScrollToErrorIsland).toHaveBeenCalled();
  });


  it('should call handle back link click handler', () => {
    renderComponent(defaultStore);
    const backLink = screen.getByText('Back');

    fireEvent(backLink, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(screen.findByText('Exporter Detaiils Page')).toBeDefined();
  });

  it('should render a help link component', () => {
    renderComponent(defaultStore);
    const helpLink = screen.getByText('Get help exporting fish from the UK (gov.uk)');
    expect(helpLink).toHaveAttribute('href', 'https://www.gov.uk/guidance/exporting-and-importing-fish-if-theres-no-brexit-deal');
  });

  it('should render a help link component with hidden text', () => {
    renderComponent(defaultStore);
    const helpSpan = screen.getByText('(opens in new tab)');
    expect(helpSpan).toBeDefined();
  });

  it('should NOT render notification banner if landings details have not been selected before the data upload feature was introduced', () => {
    renderComponent(defaultStore);
    const notificationBanner = screen.queryByText('This new page offers ways to enter products and landings for different types of export. Select an option to continue');

    expect(notificationBanner).toBeFalsy();
  });

  it('should render notification banner if landings details have been selected before the data upload feature was introduced', () => {
    renderComponent({
      config: {
        catchCertHelpUrl: 'https://www.gov.uk/guidance/exporting-and-importing-fish-if-theres-no-brexit-deal'
      },
      errors: {},
      landingsType: {
        landingsEntryOption: 'directLanding',
        generatedByContent: true
      }
    });

    const notificationBanner = screen.getByText('This new page offers ways to enter products and landings for different types of export. Select an option to continue');

    expect(notificationBanner).toBeDefined();
  });

  it('should render notification banner if redirected from copy confirm document page', () => {
    renderComponent({
      ...defaultStore,
      confirmCopyDocument: {
        copyDocumentAcknowledged: true,
        voidDocumentConfirm: false,
        documentNumber : 'XYZ-1234'
      },
    });

    const notificationBanner = screen.getByText('This draft was created by copying document XYZ-1234. You are reminded that you must not use a catch certificate or landing data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.');

    expect(notificationBanner).toBeDefined();
  });

  it('should render notification banner if redirected from copy confirm document page and void original document', () => {
    renderComponent({
      ...defaultStore,
      confirmCopyDocument: {
        copyDocumentAcknowledged: true,
        voidDocumentConfirm: true
      },
    });

    const notificationBanner = screen.getByText('This draft was created by copying a document that has now been voided. You are reminded that you must not use a catch certificate or landing data for catches that have already been exported as this is a serious offence and may result in enforcement action being taken.');

    expect(notificationBanner).toBeDefined();
  });

  it('should forbid access to the landings entry page', async () => {
    renderComponent({
      ...defaultStore,
      landingsType: {
        unauthorised: true
      }
    });

    expect(await screen.findByText('Forbidden')).toBeDefined();
  });

  it('should take a snapshot of the whole page', ()=> {
    const container = renderComponent(defaultStore);
    expect(container).toMatchSnapshot();
  });
});

describe('LandingsEntryPage, with error', () => {
  let mockOnHandleErrorClick;
  let mockWindowScrollTo;

  beforeEach(() => {
    const state = {
      config: {
        catchCertHelpUrl: 'https://www.gov.uk/guidance/exporting-and-importing-fish-if-theres-no-brexit-deal'
      },
      errors: {
        landingsEntryOptionError: 'Select a landings entry option',
        errors: [{ targetName: 'landingEntry', text: 'Select a landings entry option'}]
      },
      landingsType: {}
    };

    const store = configureStore()(state);

    const props = {
      route: {
        title: 'Create a UK catch certificate - GOV.UK',
        previousUri: '/create-catch-certificate/:documentNumber/add-exporter-details',
        nextUri: '/create-catch-certificate/:documentNumber/some-next-uri',
        path: '/create-catch-certificate/:documentNumber/landings-entry',
        journey: 'catchCertificate',
        landingsEntryOptions: null
      }
    };

    mockOnHandleErrorClick = jest.spyOn(Utils, 'onHandleErrorClick');
    mockWindowScrollTo = jest.fn()
      .mockReturnValue(undefined);

    window.scrollTo = mockWindowScrollTo;

    render(
      <Provider store={store}>
        <MemoryRouter>
          <LandingsEntryPage {...props} />
        </MemoryRouter>
      </Provider>
    );
  });

  afterEach(()=> {
    mockOnHandleErrorClick.mockRestore();
    mockWindowScrollTo.mockRestore();
  });

  it('should render a page with an error link in error island and an error within the multi choice component', () => {
    expect(screen.getAllByText('Select a landings entry option')).toHaveLength(2);
  });


  it('should call error link handler when clicked', () => {
    const errorLink = screen.getAllByText('Select a landings entry option')[0];

    fireEvent(errorLink, new MouseEvent('click', {
      bubbles: true,
      cancelable: true,
    }));

    expect(mockOnHandleErrorClick).toHaveBeenCalled();
  });

});

describe('LandingsEntryPage, load data', () => {
  const store = {
    dispatch: jest.fn()
  };

  const journey = 'catchCertificate';
  const documentNumber = 'some-document-number';

  it('will call all methods needed to load the component', () => {
    LandingsEntryPageExport.documentNumber = documentNumber;
    LandingsEntryPageExport.loadData(store, journey);

    expect(getLandingType).toHaveBeenCalled();
    expect(getLandingType).toHaveBeenCalledWith(documentNumber);
  });
});