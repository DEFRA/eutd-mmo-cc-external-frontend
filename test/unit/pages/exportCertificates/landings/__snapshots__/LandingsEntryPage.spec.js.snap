// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingsEntryPage should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <form
        action="/v1/export-certificates/landings-type?c=/create-catch-certificate/document123/landings-entry&n=/create-catch-certificate/document123/progress&documentNumber=document123"
        method="POST"
      >
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-1w80l5h-StyledColumn e4u1u2g0"
          >
            <a
              class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
              href="/create-catch-certificate/document123/add-exporter-details"
            >
              Back
            </a>
            <h1
              class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
            >
              How do you want to enter your products and landings?
            </h1>
            <div
              class="e1vcssft0 css-1phgi8s-StyledFieldset-StyledHoc e4u1u2g0"
              id="landingsEntryOption"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              />
              <fieldset
                style="margin-bottom: 30px;"
              >
                <label
                  class="css-sdrcg6-StyledHoc"
                  for="directLandingOptionEntry"
                  id="label-directLandingOptionEntry"
                >
                  <input
                    checked=""
                    class="css-1xt3i3l"
                    id="directLandingOptionEntry"
                    name="landingsEntry"
                    type="radio"
                    value="directLanding"
                  />
                  <span
                    class="css-1uwxoxa"
                  >
                    Direct Landing
                  </span>
                  <span
                    class="css-hs8umk-StyledHint-StyledHoc e4u1u2g0"
                  >
                    Recommended for UK registered fishing vessels landing and exporting their catch simultaneously in the EU (or a GB registered fishing vessel direct landing in Northern Ireland).
                  </span>
                </label>
                <label
                  class="css-sdrcg6-StyledHoc"
                  for="manualOptionEntry"
                  id="label-manualOptionEntry"
                >
                  <input
                    class="css-1xt3i3l"
                    id="manualOptionEntry"
                    name="landingsEntry"
                    type="radio"
                    value="manualEntry"
                  />
                  <span
                    class="css-1uwxoxa"
                  >
                    Manual entry
                  </span>
                  <span
                    class="css-hs8umk-StyledHint-StyledHoc e4u1u2g0"
                  >
                    Recommended for small to medium sized exports.
                  </span>
                </label>
                <label
                  class="css-sdrcg6-StyledHoc"
                  for="uploadOptionEntry"
                  id="label-uploadOptionEntry"
                >
                  <input
                    class="css-1xt3i3l"
                    id="uploadOptionEntry"
                    name="landingsEntry"
                    type="radio"
                    value="uploadEntry"
                  />
                  <span
                    class="css-1uwxoxa"
                  >
                    Upload from a CSV file
                  </span>
                  <span
                    class="css-hs8umk-StyledHint-StyledHoc e4u1u2g0"
                  >
                    Recommended for large exports. (Requires the set up of product favourites).
                  </span>
                </label>
              </fieldset>
            </div>
            <details>
              <summary
                class="govuk-details__summary"
              >
                What is a CSV file?
              </summary>
              <div
                class="govuk-details__text "
              >
                <span>
                  A CSV file is a text file that uses commas to separate its values. Each value in the file is a data field and each line is a data record.
                </span>
                <br />
                <br />
                <span>
                  Spreadsheets and some software can usually export their data to a CSV file.
                </span>
              </div>
            </details>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
          >
            Save and continue
          </button>
        </div>
      </form>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="https://www.gov.uk/guidance/exporting-and-importing-fish-if-theres-no-brexit-deal"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
